pkg/controller/ec2/securitygroup/diff.go

```diff
@@ -0,0 +1,295 @@
+/*
+Copyright 2021 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package securitygroup
+
+import (
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	awsec2 "github.com/aws/aws-sdk-go-v2/service/ec2"
+)
+
+// ruleKey represents the unique tuple (protocol, from-to port) in a
+// format supported as a map key
+type ruleKey struct {
```

Comment:
Have you had the chance to give some thoughts to my suggestion about using this as key in the map to find the rule and compare the matching rules manually by using `StringValue` and friends?

Reply:
yeah, I've tried this. It's about the same amount of code in total, but if we use lists to compare the rules manually, we have to sort them to prevent a performance degradation on large lists I. Because we don't have generics I think this makes some tricky code duplicated for each type:

```go
// for merge:

for _, r := range m.PrefixListIds {
        idx := sort.Search(len(i.prefixListIDs), func(idx int) bool {
                return prefixListIDCmp(i.prefixListIDs[idx], r) <= 0
        })

        if idx == len(i.prefixListIDs) { // nil or after last element
                i.prefixListIDs = append(i.prefixListIDs, r)
        } else if prefixListIDCmp(i.prefixListIDs[idx], r) != 0 {
                // not present, insert at idx
                i.prefixListIDs = append(i.prefixListIDs[:idx+1], i.prefixListIDs[idx:]...) // index < len(a)
                i.prefixListIDs[idx] = r
        }
}

// for diff
func (i ipPermissionMap) diffPrefixListIDs(other ipPermissionMap) []awsec2.PrefixListId {
        var ret []awsec2.PrefixListId
       for _, r := range i.prefixListIDs {
               idx := sort.Search(len(other.prefixListIDs), func(idx int) bool {
                       return prefixListIDCmp(other.prefixListIDs[idx], r) <= 0
               })
               if idx == len(other.prefixListIDs) || prefixListIDCmp(other.prefixListIDs[idx], r) != 0 {
                       ret = append(ret, r) // not found
                }
        }
        return ret

```

Later I realized there is a separate API to update descriptions only, and since CIDR etc. cannot be nil, it's  even better to rewrite as `map[cidrIPv4] -> description`, I thin, so I will do another rewrite - without any string intern magic...